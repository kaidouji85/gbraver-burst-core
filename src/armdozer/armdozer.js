// @flow

/** アームドーザID */
export type ArmDozerId = string;

// TODO バーストの種類を追加する

/** アームドーザ基本情報 */
export type Armdozer = {
  /** アームドーザID */
  id: ArmDozerId,
  /** 名前 */
  name: string,
  /** 最大HP */
  maxHp: number,
  /** 最大バッテリー */
  maxBattery: number,
  /** 攻撃 */
  power: number,
  /** スピード */
  speed: number,
  /** アームドーザの外見を指定する */
  appearance: string,
};
