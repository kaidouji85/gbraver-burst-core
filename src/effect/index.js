// @flow

import type {InputCommand} from "./input-command/input-command";
import type {Battle} from "./battle/battle";
import type {TurnChange} from "./turn-change/turn-change";
import type {StartGame} from "./start-game/start-game";
import type {BurstEffect} from "./burst/burst-effect";
import type {GameEnd} from "./game-end/game-end";
import type {BatteryDeclaration} from "./battery-declaration/battery-declaration";
import type {Reflect} from "./reflect/reflect";
import type {UpdateRemainingTurn} from "./update-remaning-turn/update-remaining-turn";

/** ゲーム中に発生する効果をまとめたもの */
export type Effect =
  StartGame |
  GameEnd |
  InputCommand |
  Reflect |
  BatteryDeclaration |
  Battle |
  TurnChange |
  BurstEffect |
  UpdateRemainingTurn;
