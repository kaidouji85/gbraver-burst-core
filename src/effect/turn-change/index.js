// @flow

import type {GameState} from "../../game-state/game-state";
import {getNextActivePlayer} from "./next-active-player";
import {getRecoveredBattery} from "./get-recovered-battery";
import type {PlayerState} from "../../game-state/player-state";

/** ターンチェンジの際に回復するバッテリー */
export const BATTERY_RECOVERY_VALUE = 3;

/**
 * ターンチェンジを実行する
 *
 * @param lastState 更新前のゲームステート
 * @return ターンチェンジ後のゲームステート
 */
export function turnChange(lastState: GameState): GameState {
  const playerIdList = lastState.players.map(v => v.playerId);
  const nextActivePlayerId = getNextActivePlayer(lastState.activePlayerId, playerIdList);
  const updatedPlayerList = lastState.players.map(player => {
    const isNextActive = player.playerId === nextActivePlayerId;
    return isNextActive ? recoverPlayerBattery(player) : player;
  });

  return {
    ...lastState,
    activePlayerId: nextActivePlayerId,
    players: updatedPlayerList,
    effect: {name: 'TurnChange'}
  };
}

/**
 * プレイヤーバッテリーを回復するヘルパー関数
 *
 * @param player 更新前のプレイヤーステート
 * @return バッテリー回復後のプレイヤーステート
 */
function recoverPlayerBattery(player: PlayerState): PlayerState {
  return {
    ...player,
    armdozer: {
      ...player.armdozer,
      battery: getRecoveredBattery(player.armdozer.battery, player.armdozer.maxBattery, BATTERY_RECOVERY_VALUE)
    }
  }
}