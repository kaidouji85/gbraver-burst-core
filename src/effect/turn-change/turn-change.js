// @flow

export type TurnChange = {
  name: 'TurnChange'
};