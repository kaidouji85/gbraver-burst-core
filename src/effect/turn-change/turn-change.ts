/** ターン変更 */
export type TurnChange = {
  name: "TurnChange";

  /** バッテリー回復量 */
  recoverBattery: number;
};
