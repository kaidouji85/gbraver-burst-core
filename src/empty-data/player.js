// @flow

import type {PlayerState} from "../state/player-state";
import {EMPTY_ARMDOZER_STATE} from "./armdozer";
import {EMPTY_PILOT_STATE} from "./pilot";

/** 空のプレイヤーステート */
export const EMPTY_PLAYER_STATE: PlayerState = {
  playerId: '',
  armdozer: EMPTY_ARMDOZER_STATE,
  pilot: EMPTY_PILOT_STATE,
};
