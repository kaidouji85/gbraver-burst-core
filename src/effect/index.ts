import { z } from "zod";

import { BatteryDeclaration, BatteryDeclarationSchema } from "./battery-declaration/battery-declaration";
import { Battle, BattleSchema } from "./battle/battle";
import { BurstEffect, BurstEffectSchema } from "./burst/burst-effect";
import { GameEnd, GameEndSchema } from "./game-end/game-end";
import { InputCommand, InputCommandSchema } from "./input-command/input-command";
import { PilotSkillEffect, PilotSkillEffectSchema } from "./pilot-skill/pilot-skill-effect";
import { Reflect, ReflectSchema } from "./reflect/reflect";
import { RightItself, RightItselfSchema } from "./right-itself/right-itself";
import { StartGame, StartGameSchema } from "./start-game/start-game";
import { TurnChange, TurnChangeSchema } from "./turn-change/turn-change";
import { UpdateRemainingTurn, UpdateRemainingTurnSchema } from "./update-remaning-turn/update-remaining-turn";

/** ゲーム中に発生する効果をまとめたもの */
export type Effect =
  | StartGame
  | GameEnd
  | InputCommand
  | Reflect
  | BatteryDeclaration
  | Battle
  | TurnChange
  | BurstEffect
  | UpdateRemainingTurn
  | RightItself
  | PilotSkillEffect;

/** Effect zodスキーマ */
export const EffectSchema = z.union([
  StartGameSchema,
  GameEndSchema,
  InputCommandSchema,
  ReflectSchema,
  BatteryDeclarationSchema,
  BattleSchema,
  TurnChangeSchema,
  BurstEffectSchema,
  UpdateRemainingTurnSchema,
  RightItselfSchema,
  PilotSkillEffectSchema,
]);