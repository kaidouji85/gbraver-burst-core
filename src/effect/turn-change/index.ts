import * as R from "ramda";

import { GameState, GameStateX } from "../../state/game-state";
import { PlayerState } from "../../state/player-state";
import { removeBatteryRecoverSkip } from "../remove-battery-recover-skip";
import { removeTurnStartBatteryCorrect } from "../remove-turn-start-battery-correct";
import { recoverBatteryOnTurnStart } from "./recover-battery-on-turn-start";
import { TurnChange } from "./turn-change";

/**
 * 次のアクティブプレイヤーステートを更新する
 * @param player プレイヤーステート
 * @param battery 回復後のバッテリー
 * @returns 更新結果
 */
const updateNextActivePlayer = (
  player: PlayerState,
  battery: number,
): PlayerState => ({
  ...player,
  armdozer: {
    ...player.armdozer,
    battery,
    effects: R.pipe(
      removeBatteryRecoverSkip,
      removeTurnStartBatteryCorrect,
    )(player.armdozer.effects),
  },
});

/**
 * ターンチェンジを実行する
 * @param lastState 更新前のゲームステート
 * @returns 実行結果、実行不可能な場合は例外を返す
 */
export function turnChange(lastState: GameState): GameStateX<TurnChange> {
  const { players, activePlayerId } = lastState;
  const nextActivePlayer = players.find((p) => p.playerId !== activePlayerId);
  if (!nextActivePlayer) {
    throw new Error("not found next active player");
  }

  const { recoverBattery, battery } =
    recoverBatteryOnTurnStart(nextActivePlayer);
  const updatedPlayers = players.map((p) =>
    p.playerId === nextActivePlayer.playerId
      ? updateNextActivePlayer(p, battery)
      : p,
  );
  return {
    ...lastState,
    activePlayerId: nextActivePlayer.playerId,
    players: updatedPlayers,
    effect: {
      name: "TurnChange",
      recoverBattery,
      reason: "TurnEnd",
    },
  };
}
