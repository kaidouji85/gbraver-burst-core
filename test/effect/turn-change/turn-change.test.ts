import path from "path";

import { ArmdozerEffect, BatteryRecoverSkip } from "../../../src";
import { turnChange } from "../../../src/effect/turn-change";
import { EMPTY_ARMDOZER_STATE } from "../../../src/empty/armdozer";
import { EMPTY_GAME_STATE } from "../../../src/empty/game-state";
import { EMPTY_PLAYER_STATE } from "../../../src/empty/player";
import type { GameState } from "../../../src/state/game-state";
import { exportSnapShotJSON, importSnapShotJSON, shouldUpdateSnapShot } from "../../snap-shot";

/** 効果 ターン開始時バッテリー回復スキップ */
const batteryRecoverSkip: BatteryRecoverSkip = {
  type: "BatteryRecoverSkip",
  period: {
    type: "Permanent",
  }
};

/** 攻撃側プレイヤー */
const attacker = {
  ...EMPTY_PLAYER_STATE,
  playerId: "attacker",
  armdozer: {
    ...EMPTY_ARMDOZER_STATE,
    battery: 2,
    maxBattery: 5,
    effects: [],
  },
};

/**
 * 防御側プレイヤーを生成する
 * @param battery 現在のバッテリー値
 * @param effects 現在の効果
 * @return 生成結果
 */
const createDefender = (battery: number, effects: ArmdozerEffect[]) => ({
  ...EMPTY_PLAYER_STATE,
  playerId: "defender",
  armdozer: {
    ...EMPTY_ARMDOZER_STATE,
    battery,
    maxBattery: 5,
    effects,
  },
});

test("ターン交代が正しく処理できる", () => {
  const defender = createDefender(2, []);
  const lastState: GameState = {
    ...EMPTY_GAME_STATE,
    players: [defender, attacker],
    activePlayerId: attacker.playerId,
  };
  const result = turnChange(lastState);
  const snapShotPath = path.join(__dirname, "turn-change.json");
  shouldUpdateSnapShot() && exportSnapShotJSON(snapShotPath, result);
  const snapShot = shouldUpdateSnapShot()
    ? result
    : importSnapShotJSON(snapShotPath);
  expect(result).toEqual(snapShot);
});

test("BatteryRecoverSkipがある場合は、バッテリー回復しない", () => {
  const defender = createDefender(2, [batteryRecoverSkip]);
  const lastState: GameState = {
    ...EMPTY_GAME_STATE,
    players: [defender, attacker],
    activePlayerId: attacker.playerId,
  };
  const result = turnChange(lastState);
  const snapShotPath = path.join(__dirname, "turn-change__battery-recover-skip.json");
  shouldUpdateSnapShot() && exportSnapShotJSON(snapShotPath, result);
  const snapShot = shouldUpdateSnapShot()
    ? result
    : importSnapShotJSON(snapShotPath);
  expect(result).toEqual(snapShot);
});