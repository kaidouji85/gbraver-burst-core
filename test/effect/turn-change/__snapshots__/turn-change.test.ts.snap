// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BatteryRecoverSkipがある場合は、バッテリー回復しない: battery-recover-skip 1`] = `
{
  "activePlayerId": "defender",
  "effect": {
    "name": "TurnChange",
    "recoverBattery": 0,
  },
  "players": [
    {
      "armdozer": {
        "battery": 2,
        "burst": {
          "recoverBattery": 0,
          "turnStartBatteryCorrect": 0,
          "type": "RecoverBattery",
        },
        "effects": [],
        "enableBurst": true,
        "hp": 3000,
        "id": "id",
        "maxBattery": 5,
        "maxHp": 3000,
        "name": "name",
        "power": 2000,
        "speed": 2000,
      },
      "pilot": {
        "enableSkill": true,
        "id": "EMPTY_PLAYER",
        "name": "名無し",
        "skill": {
          "recoverBattery": 2,
          "type": "RecoverBatterySkill",
        },
      },
      "playerId": "defender",
    },
    {
      "armdozer": {
        "battery": 2,
        "burst": {
          "recoverBattery": 0,
          "turnStartBatteryCorrect": 0,
          "type": "RecoverBattery",
        },
        "effects": [],
        "enableBurst": true,
        "hp": 3000,
        "id": "id",
        "maxBattery": 5,
        "maxHp": 3000,
        "name": "name",
        "power": 2000,
        "speed": 2000,
      },
      "pilot": {
        "enableSkill": true,
        "id": "EMPTY_PLAYER",
        "name": "名無し",
        "skill": {
          "recoverBattery": 2,
          "type": "RecoverBatterySkill",
        },
      },
      "playerId": "attacker",
    },
  ],
}
`;

exports[`ターン交代が正しく処理できる: has-no-effects 1`] = `
{
  "activePlayerId": "defender",
  "effect": {
    "name": "TurnChange",
    "recoverBattery": 3,
  },
  "players": [
    {
      "armdozer": {
        "battery": 5,
        "burst": {
          "recoverBattery": 0,
          "turnStartBatteryCorrect": 0,
          "type": "RecoverBattery",
        },
        "effects": [],
        "enableBurst": true,
        "hp": 3000,
        "id": "id",
        "maxBattery": 5,
        "maxHp": 3000,
        "name": "name",
        "power": 2000,
        "speed": 2000,
      },
      "pilot": {
        "enableSkill": true,
        "id": "EMPTY_PLAYER",
        "name": "名無し",
        "skill": {
          "recoverBattery": 2,
          "type": "RecoverBatterySkill",
        },
      },
      "playerId": "defender",
    },
    {
      "armdozer": {
        "battery": 2,
        "burst": {
          "recoverBattery": 0,
          "turnStartBatteryCorrect": 0,
          "type": "RecoverBattery",
        },
        "effects": [],
        "enableBurst": true,
        "hp": 3000,
        "id": "id",
        "maxBattery": 5,
        "maxHp": 3000,
        "name": "name",
        "power": 2000,
        "speed": 2000,
      },
      "pilot": {
        "enableSkill": true,
        "id": "EMPTY_PLAYER",
        "name": "名無し",
        "skill": {
          "recoverBattery": 2,
          "type": "RecoverBatterySkill",
        },
      },
      "playerId": "attacker",
    },
  ],
}
`;
